frame/support/procedural/src/pallet/expand/hooks.rs

```diff
@@ -25,6 +25,29 @@ pub fn expand_hooks(def: &mut Def) -> proc_macro2::TokenStream {
 	let pallet_ident = &def.pallet_struct.pallet;
 	let where_clause = &def.hooks.where_clause;
 	let frame_system = &def.frame_system;
+	let has_runtime_upgrade = def.hooks.has_runtime_upgrade;
```

Comment:
Is this true when a pallet has a `OnRuntimeUpgrade` hook or only when the storage version is lower than the pallet version?

Reply:
only when you have a function with name `on_runime_upgrade` in your `#[pallet_hooks]`. 

Perhaps I can name it `has_custom_`, because even if you don't have this, the storage version is upgraded.

Reply:
So once we add the hook we will get 

> "⚠️ running migration for {} and setting new storage version to {:?}",

forever?

Even if the current upgrade did not require a migration because no changes to the data structures were made?

Reply:
Yes, if you don't have the migration you should not have `fn on_runtime_upgrade` in your code, as it can be accidentally re-applied. 

You might say I am opinionated here, but this is exactly the scenario of how we almost burned polkadot's balance structure once, by applying a migration twice, due to forgetting to remove it from `on_runtime_upgrade`. While we have now storage versions to also prevent that, my opinion is due to such incidnets and catching them early.

Reply:
it is true when a pallet has a `OnRuntimeUpgrade` implementation (even if it is `{}`).

Reply:
Actually, storage versions are not enforced as far as I can see, further backing my argument. 

I wonder why they are not enforced? I think they can be a fantastic practice.

Reply:
But my storage migration is tied to the pallet und not to the runtime. I cannot remove it because I cannot know from which version users are upgrading. The migration checks the storage version in order to prevent double migration.

Reply:
> Actually, storage versions are not enforced as far as I can see, further backing my argument.
>
> I wonder why they are not enforced? I think they can be a fantastic practice.

what do you mean by enforcing storage versions ? storage versions are written on every runtime upgrade

Reply:
Right, I didn't think about this. But still, I think this pattern won't scale. If you want to keep the migrations in the pallet, you'd have a very big 

```
if version == 1 { /* snip */ }
else if version == 2 { /* snip */ }
else if version == 3 { /* snip */ }
else if version == 4 { /* snip */ }
```

All the way to infinity :D 

I am not sure tbh, sadly we're still learning what the optimal pattern to write migrations are and currently it is not certain yet what the optimal approach is. I'm trying to patch together an RFC about this sometime and make it standard in substrate, but currently it is all in my head and none on paper. 

@thiolliere @apopiak wdyt?

Reply:
> > Actually, storage versions are not enforced as far as I can see, further backing my argument.
> > I wonder why they are not enforced? I think they can be a fantastic practice.
> 
> what do you mean by enforcing storage versions ? storage versions are written on every runtime upgrade

Yes but we don't use them to make sure that we don't run a migration twice. We could do this: If a pallet's storage version is 3, and the crate version is also 3, even if it has a `fn migrate`, don't run it. But we don't do this. This can also be problematic in its own sense.. because the opposite is also deadly: if you forget to bump the crate version, the code is update and the storage not, and you are borked.

Reply:
For me the best practice is actually to write migration code in its own module which doesn't depend on anything in the crate, this way the migration code doesn't need to be maintained when the crate change some types.

Then either you can do something like:
```
if version <= 10 { panic!("version too old not supported") }
if version <= 11 { /* migrate from 10 to 11 */ }
if version <= 12 { /* migrate from 11 to 12 */ }
if version <= 13 { /* migrate from 12 to 13 */ }
```
or you can require user to call the migration function themself at the runtime level.

Reply:
actually I don't very like doing:
```
if version <= 10 { panic!("version too old not supported") }
if version <= 11 { /* migrate from 10 to 11 */ }
if version <= 12 { /* migrate from 11 to 12 */ }
if version <= 13 { /* migrate from 12 to 13 */ }
```

Either because it requires keeping in `Config` the old types.

For me best is to require user to call the migration function themself at the runtime level. Similar to what is done in pallet-elections-phragmen.

Reply:
Yeah I'm unsure how to resolve the conflict between keeping very long migration functions around vs not supporting upgrades from old versions. My current preferred compromise is providing this sort of full migration in the migration mod or crate and then calling whatever the pallet author thinks is best in `on_runtime_upgrade`. Then downstream users still have access to all the migrations, but you don't have to keep them around in `on_runtime_upgrade`.

Reply:
```
if version <= 10 { panic!("version too old not supported") }
if version <= 11 { /* migrate from 10 to 11 */ }
if version <= 12 { /* migrate from 11 to 12 */ }
if version <= 13 { /* migrate from 12 to 13 */ }
```

This is actually what I was planning to do. Apart from being annoying its the only thing that I can come up with that "just works". Maybe we can build some macros or functions around this pattern to make it more bearable. Because for users this is the best case. It just works.

Reply:
For now I will change the log to something that reflects it, so that we can move on: 

```
⚠️ Pallet-xxx is declaring internal migrations (which may execute) and updates storage version to xxx
```

This is because what you are doing, is neat, but what I am building is mostly for pallets that make it into polkadot and current workflow there is that before each release we want to delete all the old migrations, and such warnings can help with that.

Reply:
I understand. I am fine with that.

Reply:
updated, then can you approve? :D